### ISSUE TITLE
Vulnerable dependency gogo/protobuf (CVE-2021-3121)

### ISSUE THREAD

**Issue by Josca** [prometheus/client_golang#836]
Git versions:
- v1.9.0 (d89cf5a)

https://nvd.nist.gov/vuln/detail/CVE-2021-3121
"An issue was discovered in GoGo Protobuf before 1.3.2. plugin/unmarshal/unmarshal.go lacks certain index validation, aka the "skippy peanut butter" issue."
~~~
github.com/gogo/protobuf v1.1.1/go.mod h1:r8qH/GZQm5c6nD/R0oafs1akxWv10x8SbQlK7atdtwQ=
github.com/gogo/protobuf v1.2.0/go.mod h1:r8qH/GZQm5c6nD/R0oafs1akxWv10x8SbQlK7atdtwQ=
github.com/gogo/protobuf v1.2.1/go.mod h1:hp+jE20tsWTFYpLwKvXlhS1hjn+gTNwPg2I6zVXpSg4=
~~~
https://github.com/prometheus/client_golang/blob/d89cf5af8849fcb6d69f2c16bdf3b910e03d341e/go.sum#L76

**Comment by beorn7** [comment]
The `go.sum` file contains a lot of modules (and old versions thereof) for dependency tracking. Those lines are unlikely to ever disappear from the `go.sum` file.

However, this module doesn't use any code from github.com/gogo/protobuf, neither directly nor indirectly.

(In different news, IIUC CVE-2021-3121 correctly, the problematic part is the generated code, which you could even use without noticing in the `go.sum` file that it has anything todo with github.com/gogo/protobuf. But this module is not using any code generated with github.com/gogo/protobuf, either.)

**Comment by eero-t** [comment]
> However, this module doesn't use any code from github.com/gogo/protobuf, neither directly nor indirectly.

@beorn7 How one actually does get the list of modules that are directly or indirectly used by prometheus/client_golang?

That kind of lists are needed for license compatibility checks, security scanning etc.  Things that I think this project should already be doing...

**Comment by beorn7** [comment]
> How one actually does get the list of modules that are directly or indirectly used by prometheus/client_golang?

A good question. Go modulies is confusing in this regard. 

> Is e.g. "go mod vendor -v" or "go list -m all" output reliable list of things that would be used / linked to it?

I used to think so, but sometimes, it leads to weird results. I also got the impression things are changing with the Go version.

Security scanning of dependencies is happening via GitHub, and they seem to do a good job (as in: they have flagged actual problems in the past but haven't run into the frequent false positives that more naive security scanners run into).
